[GTK] Crash in WebCore::Cairo::drawGlyphs if threaded rendering is enabled

carlosgcampos opened the pull request:
#### b59118c2ad0b788abe2a4bd5f4163fad0d8095bd
<pre>
[GTK] Crash in WebCore::Cairo::drawGlyphs if threaded rendering is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=247628">https://bugs.webkit.org/show_bug.cgi?id=247628</a>

Reviewed by Žan Doberšek.

We are now using a single rendering thread, but the main thread can
still use FreeType during the recording phase while laying out text. So,
we can try by using a global lock for FreeType.

* Source/WebCore/platform/graphics/cairo/CairoOperations.cpp:
(WebCore::Cairo::drawGlyphs):
* Source/WebCore/platform/graphics/cairo/CairoUtilities.cpp:
(WebCore::cairoFontLock):
* Source/WebCore/platform/graphics/cairo/CairoUtilities.h:
(WebCore::CairoFtFaceLocker::CairoFtFaceLocker):
(WebCore::CairoFtFaceLocker::~CairoFtFaceLocker):

Canonical link: <a href="https://commits.webkit.org/261318@main">https://commits.webkit.org/261318@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0cfd645c5750f8376340a570731aabb866ae4768

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111106 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20246 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43671 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2533 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119950 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115062 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21619 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11348 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/2169 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116861 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16059 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99244 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103618 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98016 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30908 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44530 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12763 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32242 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86434 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13303 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9227 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18722 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15254 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4296 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0cfd645c5750f8376340a570731aabb866ae4768. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/11102

carlosgcampos committed:
[GTK] Crash in WebCore::Cairo::drawGlyphs if threaded rendering is enabled
https://bugs.webkit.org/show_bug.cgi?id=247628

Reviewed by Žan Doberšek.

We are now using a single rendering thread, but the main thread can
still use FreeType during the recording phase while laying out text. So,
we can try by using a global lock for FreeType.

* Source/WebCore/platform/graphics/cairo/CairoOperations.cpp:
(WebCore::Cairo::drawGlyphs):
* Source/WebCore/platform/graphics/cairo/CairoUtilities.cpp:
(WebCore::cairoFontLock):
* Source/WebCore/platform/graphics/cairo/CairoUtilities.h:
(WebCore::CairoFtFaceLocker::CairoFtFaceLocker):
(WebCore::CairoFtFaceLocker::~CairoFtFaceLocker):

Canonical link: https://commits.webkit.org/261318@main

webkit-commit-queue commented:
Committed 261318@main (b59118c2ad0b): <https://commits.webkit.org/261318@main>

Reviewed commits have been landed. Closing PR #11102 and removing active labels.
